tests/pds/dataservice_test.go

```diff
@@ -1238,67 +1251,73 @@ func UpgradeDataService(dataservice, oldVersion, oldImage, dsVersion, dsBuild st
+		ValidateDeployments(resourceTemp, storageOp, config, int(replicas), dataServiceVersionBuildMap)
+	})
 
-		Step("Running Workloads before scaling up of dataservices ", func() {
-			var params pdslib.WorkloadGenerationParams
+	Step("Running Workloads before scaling up of dataservices ", func() {
+		var params pdslib.WorkloadGenerationParams
+		if Contains(dataServicePodWorkloads, ds.Name) || Contains(dataServiceDeploymentWorkloads, ds.Name) {
 			pod, dep, err = RunWorkloads(params, ds, deployment, namespace)
 			log.FailOnError(err, "Error while genearating workloads")
-		})
+		}
 
-		defer func() {
-			Step("Delete the workload generating deployments", func() {
-				if Contains(dataServiceDeploymentWorkloads, ds.Name) {
-					log.InfoD("Deleting Workload Generating pods %v ", dep.Name)
-					err = pdslib.DeleteK8sDeployments(dep.Name, namespace)
-				} else if Contains(dataServicePodWorkloads, ds.Name) {
-					log.InfoD("Deleting Workload Generating pods %v ", pod.Name)
-					err = pdslib.DeleteK8sPods(pod.Name, namespace)
-				}
-				log.FailOnError(err, "error deleting workload generating pods for ds %s", dataservice)
-			})
-		}()
+	})
+
+	defer func() {
+		Step("Delete the workload generating deployments", func() {
+			if Contains(dataServiceDeploymentWorkloads, ds.Name) {
+				log.InfoD("Deleting Workload Generating pods %v ", dep.Name)
+				err = pdslib.DeleteK8sDeployments(dep.Name, namespace)
+			} else if Contains(dataServicePodWorkloads, ds.Name) {
+				log.InfoD("Deleting Workload Generating pods %v ", pod.Name)
+				err = pdslib.DeleteK8sPods(pod.Name, namespace)
+			}
+			log.FailOnError(err, "error deleting workload generating pods for ds %s", dataservice)
+		})
+	}()
 
-		Step("Update the data service patch versions", func() {
-			log.Infof("Version/Build: %v %v", dsVersion, dsBuild)
-			updatedDeployment, err := pdslib.UpdateDataServiceVerison(deployment.GetDataServiceId(), deployment.GetId(),
-				dataServiceDefaultAppConfigID,
-				replicas, dataServiceDefaultResourceTemplateID, dsBuild, namespace, dsVersion)
-			log.FailOnError(err, "Error while updating data services")
-			log.InfoD("data service deployed %v ", dataservice)
+	Step("Update the data service patch versions", func() {
+		log.Infof("Version/Build: %v %v", dsVersion, dsBuild)
+		updatedDeployment, err := pdslib.UpdateDataServiceVerison(deployment.GetDataServiceId(), deployment.GetId(),
+			dataServiceDefaultAppConfigID,
+			replicas, dataServiceDefaultResourceTemplateID, dsBuild, dsVersion)
+		log.FailOnError(err, "Error while updating data services")
+		log.InfoD("data service deployed %v ", dataservice)
 
-			resourceTemp, storageOp, config, err := pdslib.ValidateDataServiceVolumes(updatedDeployment, dataservice, dataServiceDefaultResourceTemplateID, storageTemplateID, namespace)
-			log.FailOnError(err, "error on ValidateDataServiceVolumes method")
+		err = pdslib.ValidateDataServiceDeployment(updatedDeployment, namespace)
```

Comment:
For now it's ok, but it's becoming a big mess soon. We should not be incorporating logic in test case file. Let's soon re-design our test case writing to just have a absolutely neccessary number of calls to functions that does all the magic and make it extremely modular. Plus let's remove all functions from test case file as well. 
@gauraa2, @Madanagopal19  & @jkagliwal-px : Let's sit in a meeting and re-design the tests. This model will cause our test case files to go beyond tens of thousands of lines soon - making the tests totally unmanageable. 

^ This is a generic comment not specific to this change. For now it is ok, but we will change it soon.

Reply:
sure

Reply:
I had same concern ! Lets re-design.